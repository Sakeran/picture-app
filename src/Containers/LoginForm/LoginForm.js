import React from 'react';
import PropTypes from 'prop-types';

import Formsy from 'formsy-react';
import TextInput from '../formComponents/TextInput';

import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';

class LoginForm extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      canSubmit: false
    };
  }

  enableButton = () => {
    this.setState({
      canSubmit: true
    });
  }

  disableButton = () => {
    this.setState({
      canSubmit: false
    })
  }

  submit = ({username, password}) => {
    this.props.sendFunc(username, password)
    .then(userData => {
      console.log(userData);
      const user = JSON.parse(userData);
      if (!user) {
        console.warn('Login Attempt returned null');
        return;
      }
      this.props.loginUser(user);
    });
  }

  render() {
    if (this.props.user) {
      return <Redirect to="/" />
    }
    return (
      <Formsy.Form onValidSubmit={this.submit}
                   onValid={this.enableButton}
                   onInvalid={this.disableButton}>
        <TextInput name="username"
                   title="Username"
                   required
                   validations={{
                     minLength: 1,
                     hasSpaces: (values, value) =>
                                (value.indexOf(" ") === -1)
                   }}
                   validationErrors={{
                     minLength: 'Username cannot be blank. ',
                     hasSpaces: 'Username cannot contain spaces'
                   }}
                   value="" />
        <TextInput name="password"
                   type="password"
                   title="Password"
                   required
                   validations="minLength:1"
                   validationError="Invalid Password"
                   value="" />
        <input type="submit"
                className="button"
                disabled={!this.state.canSubmit}
                value="Log In" />
      </Formsy.Form>
    );
  }
}

LoginForm.propTypes = {
  sendFunc: PropTypes.func.isRequired,
  loginUser: PropTypes.func,
  user: PropTypes.object
};

const mapStateToProps = (state) => ({
  user: state.common.user
});

const mapDispatchToProps = (dispatch) => ({
  loginUser: (user) => dispatch({type: 'LOGIN_USER', user: user})
});

const ConnectedLoginForm = connect(mapStateToProps, mapDispatchToProps)(LoginForm);

export { LoginForm };
export default ConnectedLoginForm;
