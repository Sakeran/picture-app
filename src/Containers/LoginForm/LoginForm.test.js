import React from 'react';
import { LoginForm } from './LoginForm';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

test('renders all fields correctly', () => {
  const fn = () => null;
  const component = mount(<LoginForm sendFunc={fn}/>);
  expect(component.find('input[name="username"]').exists()).toBe(true);
  expect(component.find('input[name="password"]').exists()).toBe(true);
  expect(component.find('input[type="submit"]').exists()).toBe(true);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});

test('calls a login function when given valid data', (done) => {
  const userData = {username: 'user'};
  const sendFn = () => Promise.resolve(JSON.stringify(userData));
  const loginFn = jest.fn();
  const component = mount(<LoginForm sendFunc={sendFn} loginUser={loginFn} />);

  component.find('input[name="username"]')
  .simulate('change', { target: {value: 'user'} });

  component.find('input[name="password"]')
  .simulate('change', { target: {value: 'password'} });

  component.find('form').simulate('submit');

  process.nextTick(() => {
    expect(loginFn.mock.calls.length).toBe(1);
    done();
  });
});

test('does not call a login function when given invalid data', (done) => {
  const userData = {username: 'user'};
  const sendFn = () => Promise.resolve(JSON.stringify(userData));
  const loginFn = jest.fn();
  const component = mount(<LoginForm sendFunc={sendFn} loginUser={loginFn} />);

  component.find('input[name="username"]')
  .simulate('change', { target: {value: 'user'} });

  // Skip password field on purpose

  component.find('form').simulate('submit');

  process.nextTick(() => {
    expect(loginFn.mock.calls.length).toBe(0);
    done();
  });
});
