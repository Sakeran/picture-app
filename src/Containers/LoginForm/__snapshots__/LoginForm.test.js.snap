// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders all fields correctly 1`] = `
<LoginForm
  sendFunc={[Function]}
>
  <Formsy
    onChange={[Function]}
    onError={[Function]}
    onInvalid={[Function]}
    onInvalidSubmit={[Function]}
    onSubmit={[Function]}
    onSuccess={[Function]}
    onValid={[Function]}
    onValidSubmit={[Function]}
    preventExternalInvalidation={false}
    validationErrors={null}
  >
    <form
      onSubmit={[Function]}
    >
      <Formsy(TextInput)
        name="username"
        required={true}
        title="Username"
        validationError=""
        validationErrors={
          Object {
            "hasSpaces": "Username cannot contain spaces",
            "minLength": "Username cannot be blank. ",
          }
        }
        validations={
          Object {
            "hasSpaces": [Function],
            "minLength": 1,
          }
        }
        value=""
      >
        <TextInput
          getErrorMessage={[Function]}
          getErrorMessages={[Function]}
          getValue={[Function]}
          hasValue={[Function]}
          isFormDisabled={[Function]}
          isFormSubmitted={[Function]}
          isPristine={[Function]}
          isRequired={[Function]}
          isValid={[Function]}
          isValidValue={[Function]}
          name="username"
          required={true}
          resetValue={[Function]}
          setValidations={[Function]}
          setValue={[Function]}
          showError={[Function]}
          showRequired={[Function]}
          title="Username"
          validationError=""
          validationErrors={
            Object {
              "hasSpaces": "Username cannot contain spaces",
              "minLength": "Username cannot be blank. ",
            }
          }
          validations={
            Object {
              "hasSpaces": [Function],
              "minLength": 1,
            }
          }
          value=""
        >
          <div
            className="field"
          >
            <label
              htmlFor="username"
            >
              Username
            </label>
            <input
              name="username"
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="field-error"
            />
          </div>
        </TextInput>
      </Formsy(TextInput)>
      <Formsy(TextInput)
        name="password"
        required={true}
        title="Password"
        type="password"
        validationError="Invalid Password"
        validationErrors={Object {}}
        validations="minLength:1"
        value=""
      >
        <TextInput
          getErrorMessage={[Function]}
          getErrorMessages={[Function]}
          getValue={[Function]}
          hasValue={[Function]}
          isFormDisabled={[Function]}
          isFormSubmitted={[Function]}
          isPristine={[Function]}
          isRequired={[Function]}
          isValid={[Function]}
          isValidValue={[Function]}
          name="password"
          required={true}
          resetValue={[Function]}
          setValidations={[Function]}
          setValue={[Function]}
          showError={[Function]}
          showRequired={[Function]}
          title="Password"
          type="password"
          validationError="Invalid Password"
          validationErrors={Object {}}
          validations="minLength:1"
          value=""
        >
          <div
            className="field"
          >
            <label
              htmlFor="password"
            >
              Password
            </label>
            <input
              name="password"
              onChange={[Function]}
              type="password"
              value=""
            />
            <span
              className="field-error"
            />
          </div>
        </TextInput>
      </Formsy(TextInput)>
      <input
        className="button"
        disabled={true}
        type="submit"
        value="Log In"
      />
    </form>
  </Formsy>
</LoginForm>
`;
